packages/collector/test/metrics/appWithWorkerThread/test.js

```diff
@@ -0,0 +1,120 @@
+const supportedVersion = require('@instana/core').tracing.supportedVersion;
+const globalAgent = require('../../globalAgent');
+
+// See https://instana.kanbanize.com/ctrl_board/56/cards/48699/details/ for more details on worker threads.
+
+const mochaSuiteFn =
+  supportedVersion(process.versions.node) &&
+  // worker threads became stable in Node.js 12.
+  semver.gte(process.versions.node, '12.0.0')
+    ? describe
+    : describe.skip;
+
+mochaSuiteFn('worker threads', function () {
+  this.timeout(config.getTestTimeout());
+
+  globalAgent.setUpCleanUpHooks();
+  const agentControls = globalAgent.instance;
+
+  describe('with in-app require main thread but not in worker thread', () => {
+    const controls = createProcess();
+
+    it('will neither report metrics nor spans from a worker thread', async () => {
+      await verify({
+        controls,
+        expectSpans: false
+      });
+    });
+  });
+
+  describe('with in-app require in main thread and in worker thread', () => {
+    const controls = createProcess({
+      REQUIRE_INSTANA_IN_WORKER_THREAD: true
+    });
+
+    it('must not report metrics from a worker thread', async () => {
+      await verify({
+        controls,
+        expectSpans: true
+      });
+    });
+  });
+
+  describe('with NODE_OPTIONS/pre-require', () => {
+    const controls = createProcess({
+      NODE_OPTIONS: '--require ../../../src/immediate'
+    });
+
+    it('must not report metrics from a worker thread', async () => {
+      await verify({
+        controls,
+        expectSpans: true
+      });
+    });
+  });
+
+  function createProcess(env = {}) {
+    env.WORKER_THREAD_APP_PORT = 3216;
+    return new ProcessControls({
+      appPath: path.join(__dirname, 'app'),
+      cwd: __dirname,
+      useGlobalAgent: true,
+      env
+    }).registerTestHooks();
+  }
+
+  async function verify({ controls, expectSpans }) {
+    let allMetrics;
+    let allNames;
+
+    // First, make sure that the announce from the main thread has happened by retrying until we find
+    //   (a) a metric object that has a "name" attribute, and
+    //   (b) where the value for "name" is the expected application name (from the main thread).
+    await retry(async () => {
+      allMetrics = await agentControls.getAllMetrics(controls.getPid());
+      allNames = [];
+      allMetrics.forEach(metrics => {
+        if (metrics.data.name) {
+          allNames.push(metrics.data.name);
+        }
+      });
+      expect(allNames).to.contain('worker-thread-test-app');
+    });
+
+    // We have made sure that the main thread has announced itself. Now we give the worker thread a chance to announce
+    // itself and send metrics. Actually, we do _not_ want the worker thread to send metrics on its own, but to verify
+    // that this does not happen, we need to wait a bit so that the test actually would fail if the worker thread sent
+    // metrics.
+    await delay(800);
```

Comment:
I was wondering if this can create another flaky test.
Is there a way to determine if the worker thread has collected the spans yet? (I have not looked for a solution)

Reply:
Right, the section for checking if spans have been produced needs to be wrapped in another `retry` block. Fixed.